fuzz/fuzz_targets/cranelift-fuzzgen.rs

```diff
@@ -359,6 +359,14 @@ fuzz_target!(|testcase: TestCase| {
     // This is the default, but we should ensure that it wasn't accidentally turned off anywhere.
     assert!(testcase.isa.flags().enable_verifier());
 
+    let mut ctrl_planes = testcase.ctrl_planes.clone();
+    if let Some(fuel) =
+        std::env::args().find_map(|arg| arg.strip_prefix("--fuel=").map(|s| s.to_owned()))
```

Comment:
This is unfortunately not the right way to go about this, I think: we should funnel all settings that control compilation through the settings plumbing, rather than pulling inputs from implicit/ambient environment. (The latter may seem easier for a specific setting, such as running a fuzz case on the command line, but can lead to really surprising and subtle bugs later, and also precludes other options like a tool that bisects automatically.)

I think this is a good opportunity to flesh out the text format, not just for printing but for parsing as well. If we can print a textual form of the control-plane (analogous to the `set option=value`, we could have `control_plane chaos=... fuel=...` or somesuch) and parse it, then the fuzzing debug flow would be: print the formatted fuzz testcase as CLIF; then iterate on it by tweaking the setting. Alternately we can add command-line options to `clif-util test`.

Reply:
I'm assuming the subtle bugs you're referring to would be forgetting to set the fuel correctly in some case. To address this, I think it's necessary to remove the `Arbitrary` implementation of the control plane to avoid constructing one without having consulted the fuel setting first.

The ad-hoc CLI argument parsing is still there, now in the TestCase arbitrary implementation. Which feels a little dirty. But I didn't see another way, I thought it was more important to ensure a control plane is created with the fuel already specified.

The text representation and parsing I'm still working on.

If the control plane fuel is part of cranelift's settings, should it be part of the text representation of a control plane as well? How should a text representation like this be interpreted:
```
; <other_flags>
set control_plane_fuel=16
; <target>

control_plane fuel=8 data=1100
; <function>
```
What if the control plane value (8) is a different one than the cranelift setting (16)?

Reply:
> I'm assuming the subtle bugs you're referring to would be forgetting to set the fuel correctly in some case. To address this, I think it's necessary to remove the Arbitrary implementation of the control plane to avoid constructing one without having consulted the fuel setting first.

Is it necessarily only one or the other? Why is discarding `Arbitrary` altogether the only other choice? For example, could we instead inject a fuel value into the control-plane object after creating one with `Arbitrary`?

More generally: reading environment variables from inside a library is to be avoided because it circumvents the embedder's control of the library, and means that the library's behavior will be non-deterministic from the point of view of the API user. I ask for a control-plane with some bytes and compile on my system; you do the same on your system; we get two different results, because our ambient process state is different.

So the sort of bug I worry about isn't "forgot to set fuel" but "perplexing divergence in behavior that only makes sense once we check all basic assumptions and examine our shell configuration for environment variables" -- much less fun to debug!

Reply:
The argument parsing is still in the same file, the fuzz target
`cranelift-fuzzgen`. Which library do you mean?

I guess the problem I was working around is that the flags are generated in the `Arbitrary` implementation of `TestCase`, which doesn't take any arguments. And it's not possible to change the flags afterwards (?).

So it seems to me the CLI argument must be parsed before or during the the construction of an arbitrary `TestCase`. Otherwise we can't set the appropriate cranelift flag. Parsing CLI args *before* test case generation with arbitrary doesn't work though:
```rust

fuzz_target!(|data: &[u8]| {
    let fuel = std::env::args()
        .find_map(|arg| arg.strip_prefix("--fuel=").map(|s| s.to_owned()))
        .map(|fuel| fuel.parse().expect("fuel should be a valid integer"));

    let mut testcase = TestCase::generate(&mut Unstructured::new(data), fuel).unwrap();
    for i in 0..testcase.ctrl_planes.len() {
        testcase.ctrl_planes[i].set_fuel(testcase.isa.flags().control_plane_fuel())
    }
```
The problem here is `TestCase::generate(/* ... */).unwrap()`. As far as I can tell from the cargo fuzz book and the expansion of the `fuzz_target!` macro, the body of a fuzz target is not allowed to reject an input (`data: &[u8]`) because its structure is unsuitable. That should (can?) only be done by putting the structured input right as the argument (fuzz_target!(|data: TestCase|). I might be wrong about this.

So that leads me to think our only option is to parse the CLI argument in some implementation of `Arbitrary`. Would it be better to make a separate struct and `Arbitrary` implementation just for the purpose of parsing the CLI args, to separate these concerns? Pseudo-code:
```rust
struct Foo(TestCase);

impl<'a> Arbitrary<'a> for Foo {
    fn arbitrary(u: &mut Unstructured<'a>) -> arbitrary::Result<Self> {
        let fuel = parse_fuel_from_cli_args();
        let mut testcase = TestCase::generate(u, fuel)?;
        testcase.set_all_fuel(testcase.isa.flags().control_plane_fuel());
        Ok(Foo(testcase))
    }
}

fuzz_target!(|foo: Foo| {
    let Foo(testcase) = foo;
```

Reply:
> The argument parsing is still in the same file, the fuzz target `cranelift-fuzzgen`. Which library do you mean?

I guess I meant it a little more abstractly: the implementation of `Arbitrary` is a library layer of sorts, and should depend only on its inputs.

> I guess the problem I was working around is that the flags are generated in the `Arbitrary` implementation of `TestCase`, which doesn't take any arguments. And it's not possible to change the flags afterwards (?).

That I think is the key assumption that is guiding this into the direction I'm not sure about :-) The `TestCase` should be defined to contain only the plain-old-data inputs; we shouldn't be building the compiler context with flags inside the `Arbitrary` impl itself. We should be able to then build whatever context we need in the fuzzing entrypoint: the compiler context is a pure function of the `TestCase` and the fuel parameter.

So I'm imagining something like:

```rust
pub struct TestCase {
    flags: cranelift_codegen::Settings,
    // ...
}
```

and then

```rust
impl TestCase {
    fn build_isa(&self, fuel: ...) -> isa::OwnedTargetIsa { ... }
}
```

or even `fn build_isa(&self, user_params: &UserParams)` and build the user params however we want (e.g. `UserParams::from_env()`). That makes the plumbing explicit and removes the dependency on environment variables from inside what should be a pure function :-)

----

Taking a step back, I think we've discussed at least two different kinds of "fuel" and three different sources of the setting, and I want to at least outline this taxonomy here to make sure we're on the same page and have a framework to think about it further.

Kinds of fuel:
- Fuel that "masks" other settings (e.g. chaos-decision values), and
- Fuel that is a separate dimension or axis that controls many different mutations in the compiler.

Overall the latter is the more useful long-term, but the former is useful when controlled by the user to bisect failures, so both are interesting in some context.

Sources of fuel setting:
- User parameter (CLIF option on clif-util, environment variable read from toplevel fuzz-test entry point, ...). Useful for the bisection kind of fuel.
- Cranelift settings, or a separate control-plane setting that is passed in alongside settings. Useful for the kind of fuel that can be used to limit optimization effort (the latter), and also for bisection when driven in certain ways (CLIF file editing etc).
- Programmatic generation of control-plane state: useful for e.g. tools that try to bisect automatically.

The approach we're trying to take here is squarely the first kind of fuel, for now, and with the first source. If we can fix up the plumbing a little bit as above to isolate the read-of-ambient-state in as clean a way as possible, then I think I'm happy. Thanks!